server/conn.go

```diff
@@ -1706,23 +1706,20 @@ func (cc *clientConn) handleQuery(ctx context.Context, sql string) (err error) {
 		}
 		retryable, err = cc.handleStmt(ctx, stmt, parserWarns, i == len(stmts)-1)
 		if err != nil {
+			if !retryable || !errors.ErrorEqual(err, storeerr.ErrTiFlashServerTimeout) {
+				break
+			}
 			_, allowTiFlashFallback := cc.ctx.GetSessionVars().AllowFallbackToTiKV[kv.TiFlash]
-			if allowTiFlashFallback && errors.ErrorEqual(err, storeerr.ErrTiFlashServerTimeout) && retryable {
+			if allowTiFlashFallback {
 				// When the TiFlash server seems down, we append a warning to remind the user to check the status of the TiFlash
 				// server and fallback to TiKV.
 				warns := append(parserWarns, stmtctx.SQLWarn{Level: stmtctx.WarnLevelError, Err: err})
-				func() {
-					delete(cc.ctx.GetSessionVars().IsolationReadEngines, kv.TiFlash)
-					defer func() {
-						cc.ctx.GetSessionVars().IsolationReadEngines[kv.TiFlash] = struct{}{}
-					}()
-					_, err = cc.handleStmt(ctx, stmt, warns, i == len(stmts)-1)
-				}()
+				delete(cc.ctx.GetSessionVars().IsolationReadEngines, kv.TiFlash)
+				_, err = cc.handleStmt(ctx, stmt, warns, i == len(stmts)-1)
+				cc.ctx.GetSessionVars().IsolationReadEngines[kv.TiFlash] = struct{}{}
```

Comment:
Here we use `defer` because we should add `TiFlash` back to `IsolationReadEngines` if `handleStmt` panics. Check https://github.com/pingcap/tidb/pull/22459#discussion_r565106002. However, It seems the connection will be closed if `handleStmt` panics so maybe adding `TiFlash` back to `IsolationReadEngines` is unnecessary. @qw4990

Reply:
@xuyifangreeneyes Oh there's such a hidden logic here...
I think this panic should not be recovered by adding back TiFlash as a valid `IsolationReadEngines`.

Reply:
@dragonly Can you remove similar `defer` in another place `(*clientConn).handleStmtExecute`, thx~